### PR TITLE
Python Node Editor Visual Update - part #2

### PR DESCRIPTION
**Pull request by dnenov**
### Purpose

This PR is a follow-up and is based on https://github.com/DynamoDS/Dynamo/pull/13732. It goes further in customizing the Python Script editor in a number of ways.

#### Exit saved/unsaved script
exiting_with_unsaved_changes

#### Folding tests
folding_test

### Declarations

Check these if you believe they are true

- [x] The codebase is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [x] User facing strings, if any, are extracted into `*.resx` files
- [x] All tests pass using the self-service CI.
- [x] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated

### Release Notes

- icon update
- introduced zoom-in/out buttons
- introduced undo/redo buttons
- change multiple reserved keyword colors
- enabled text-folding
- custom (tab) indentation strategy 
- now prompts the user to save unsaved changes on 'Esc' exit

### Reviewers

@sm6srw
@Amoursol

### FYIs

@Jingyi-Wen
@hwahlstrom

### CODE REVIEW

File: src/Libraries/PythonNodeModelsWpf/PythonIndentationStrategy.cs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+using System.Text;
+using System.Threading.Tasks;
+
+namespace PythonNodeModelsWpf
+{
+    /// <summary>
+    /// Custom Indentation Strategy for Python 
+    /// https://csharp.hotexamples.com/site/file?hash=0x1abeea836b72c6db1910df1767e84d7bfcb620ed58a499d176fdf158851c1d5f&fullName=Yanitta/LuaIndentationStrategy.cs&project=Konctantin/Yanitta
+    /// </summary>
+    public class PythonIndentationStrategy : DefaultIndentationStrategy
```
Comment (mjkkirschner) [inline src/Libraries/PythonNodeModelsWpf/PythonIndentationStrategy.cs:16]: can this be made internal?

File: src/Libraries/PythonNodeModelsWpf/TabFoldingStrategy.cs [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+using System.Text;
+using System.Windows.Media.Animation;
+
+namespace PythonNodeModelsWpf
+{
+    /// <summary>
+    /// Allows producing tab based foldings
+    /// https://stackoverflow.com/questions/47224064/avalonedit-foldingstrategy-by-indent-python/47577500#47577500
+    /// </summary>
+    public class TabFoldingStrategy : AbstractFoldingStrategy
```
Comment (mjkkirschner) [inline src/Libraries/PythonNodeModelsWpf/TabFoldingStrategy.cs:14]: internal?

File: src/Libraries/PythonNodeModelsWpf/PythonIndentationStrategy.cs (hunk 1 above)
Comment (dnenov) [reply src/Libraries/PythonNodeModelsWpf/PythonIndentationStrategy.cs:16]: Oh, yes, absolutely! Thank you for spotting that and the other one out, I will change them to internal.

File: src/Libraries/PythonNodeModelsWpf/TabFoldingStrategy.cs (hunk 2 above)
Comment (dnenov) [reply src/Libraries/PythonNodeModelsWpf/TabFoldingStrategy.cs:14]: Yup will do that! This one is still very finicky, still working on making it more robust. It requires an update run virtually on each keystroke. I actually wanted to perhaps seek advice from you on how to profile performance with and without folding. Can I ask in Slack?

**Comment by Amoursol** [comment]
@dnenov can you please also put an image or GIF inside of this PR? We often refer back to these when creating presentations etc 😊

**Comment by dnenov** [comment]
> @dnenov can you please also put an image or GIF inside of this PR? We often refer back to these when creating presentations etc 😊

Of course! Adding them in as we are closing these chapters.

**Comment by sm6srw** [comment]
Tests passed here: https://master-5.jenkins.autodesk.com/job/Dynamo/job/DynamoSelfServe/job/pullRequestValidation/8583/

**Review by sm6srw — APPROVED** [review]

File: src/Libraries/PythonNodeModelsWpf/ScriptEditorWindow.xaml [hunk 3]
```diff
@@ -121,6 +121,71 @@
...
+                                    BorderBrush="#B8ABA1"
+                                    CornerRadius="2"
+                                    BorderThickness="1">
+                                <Grid x:Name="inner"
+                                      Background="Transparent">
+                                    <TextBlock x:Name="text"
+                                               HorizontalAlignment="Center"
+                                               VerticalAlignment="Center"
+                                               FontSize="14px"
+                                               Foreground="#3C3C3C"
```
Comment (reddyashish) [inline src/Libraries/PythonNodeModelsWpf/ScriptEditorWindow.xaml:144]: Use the color variables from Dynamo brushes files, wherever possible.

File: src/Libraries/PythonNodeModelsWpf/ScriptEditorWindow.xaml [hunk 4]
```diff
@@ -312,6 +435,33 @@
...
+                        <ColumnDefinition Width="44"/>
+                        <ColumnDefinition Width="*"/>
+                        <ColumnDefinition Width="180"/>
+                    </Grid.ColumnDefinitions>
+                        <Border Background="#FAA21B" Grid.Column="0"> 
+                            <Image Source="/PythonNodeModelsWpf;component/Resources/alert.png" Width="24" Height="24"/>
+                        </Border>
+                        <StackPanel Orientation="Vertical" VerticalAlignment="Center" Grid.Column="1">
+                        <TextBlock Margin="15 5" FontSize="14"  FontFamily="{StaticResource ArtifaktElementBold}" Text="Are you sure you want to leave?"/>
+                        <TextBlock Margin="15 5" FontSize="14" FontFamily="{StaticResource ArtifaktElementRegular}" TextWrapping="WrapWithOverflow" Text="It looks like you are still working on this. You may lose any unsaved changes if you leave now"/>
```
Comment (reddyashish) [inline src/Libraries/PythonNodeModelsWpf/ScriptEditorWindow.xaml:456]: This text would need localization, can you add it to resources file.

File: src/Libraries/PythonNodeModels/Properties/Resources.Designer.cs [hunk 5]
```diff
@@ -205,7 +205,7 @@ public class Resources {
         }
         
         /// <summary>
-        ///   Looks up a localized string similar to Learn about Python3 updates..
+        ///   Looks up a localized string similar to Learn about the Python 2 to Python 3 conversion.
```
[the remark this answers is not in the document]
Comment (reddyashish) [reply src/Libraries/PythonNodeModels/Properties/Resources.Designer.cs:208]: yeah looks like the resources were changed in the previous PR. sounds good as long as it is consistent.

**Comment by reddyashish** [comment]
@dnenov There is one conflict on this PR. PTAL.

File: src/Libraries/PythonNodeModelsWpf/ScriptEditorWindow.xaml (hunk 3 above)
Comment (dnenov) [reply src/Libraries/PythonNodeModelsWpf/ScriptEditorWindow.xaml]: Resolved

File: src/Libraries/PythonNodeModelsWpf/ScriptEditorWindow.xaml (hunk 4 above)
Comment (dnenov) [reply src/Libraries/PythonNodeModelsWpf/ScriptEditorWindow.xaml]: Resolved, thank you! It slipped under the radar.

**Review by reddyashish — APPROVED** [review]
LGTM once the conflict is resolved

**Comment by reddyashish** [comment]
@dnenov Thank you for addressing everything. Will wait for the self-service to finish.

**Comment by reddyashish** [comment]
Passed, merging this.

**Pull request merged**